services/alerts/src/AlertsProvider.tsx

```diff
@@ -0,0 +1,20 @@
+import React, { useState, useMemo } from 'react'
+import { AlertsContext } from './AlertsContext'
+import { AlertsManagerContext, makeAlertManager } from './AlertsManagerContext'
+import { Alert, AlertsManager } from './types'
+
+export const AlertsProvider = ({ children }: { children: React.ReactNode }) => {
+    const [alerts, setAlerts] = useState<Alert[]>([])
+    const alertsManager: AlertsManager = useMemo(
+        () => makeAlertManager(setAlerts),
+        []
+    )
```

Comment:
This part used to be using the `useState` hook, but `useMemo` seemed to be a better fit with what we want to do.

Reply:
I don't think so - `useMemo` is a performance optimization which [does not provide a semantic guarantee](https://reactjs.org/docs/hooks-reference.html#usememo).  If `useMemo` is re-initialized here it will re-render every context consumer.  Not technically the end of the world, as this should be rare, but probably better to use State I think?

Reply:
If I interpret this correctly, this will never cause a re-render, so we could just use `useRef`

> It’s handy for keeping any mutable value around similar to how you’d use instance fields in classes.
(From the [reactjs.org docs](https://reactjs.org/docs/hooks-reference.html#useref))

[Is there something like instance variables?](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables)

> The useRef() Hook isn’t just for DOM refs. The “ref” object is a generic container whose current property is mutable and can hold any value, similar to an instance property on a class.

Reply:
I read the docs section about `useMemo` not providing a semantic guarantee and you're absolute correct in saying that `useMemo` is not what I should have gone with here.

> If `useMemo` is re-initialized here it will re-render every context consumer. Not technically the end of the world, as this should be rare

In its current form, this is actually a bigger deal than you might expect. When `makeAlertManager` is called, the  `id` is being reinitialised as `0`. This means we end up with duplicate ids in the alerts list. Finding a way to ensure `makeAlertManager` is called just once from the `AlertsProvider` would be my preferred way of preventing this `id` issue.

I still have a lot of learning to do reg. hooks, so perhaps my reasoning is not correct, but I was thinking:
- `useMemo` isn't a good option here because of it not providing a semantic guarantee.
- I can see how the previous version based on `useState` would work correctly, but to me it sort of feels like we are not using it as it was intended to, because we are not intending to update this `alertManager`, and we certainly don't want the UI to respond to changes in the `alertManager` either. If this were a class based component I would have gone with `this.alertManager`, not `this.state.alertManager` (an instance variable not a state variable)....
- I've seen `useRef` being used as "functional equivalent of a class instance variable", and [this article](https://medium.com/trabe/react-useref-hook-b6c9d39e2022) and [the official docs](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables) seem to suggest this is the hook to use for our use case.

So based on my pretty limited knowledge I'd say we should probably go with `useRef` and set it in a useEffect hook. A bit like this:
```javascript
export const AlertsProvider = ({ children }: { children: React.ReactNode }) => {
    const alertsManagerRef = useRef()
    const [alerts, setAlerts] = useState<Alert[]>([])

    useEffect(() => {
        alertsManagerRef.current = makeAlertManager(setAlerts)
    }, [])

    return (
        <AlertsManagerContext.Provider value={alertsManagerRef.current}>
            <AlertsContext.Provider value={alerts}>
                {children}
            </AlertsContext.Provider>
        </AlertsManagerContext.Provider>
    )
}
```

When I was writing the above in my editor, I got this TypeScript error: `Type 'AlertsManager' is not assignable to type 'undefined'`. I haven't looked into this issue yet, but it could be a valid problem where `alertsManagerRef.current` is `undefined` during the first render. We probably don't want that, and I guess this would justify the (mis)use of `useState` here.... On the other hand, it could just be an issue with how I've declared the types...

Assuming we are just dealing with a type issue here, would you agree that the `useRef` hook is the most correct one to use in this case? @Mohammer5 @amcgee

Reply:
See my message above, I'd also use `useRef`.

> When I was writing the above in my editor, I got this TypeScript error: Type 'AlertsManager' is not assignable to type 'undefined'.

I assume that this is because you're not providing a type to `useRef`, I guess you'll have to do

```js
const allertManagerRef = useRef<AlertsManager>(defaultAlertsManager)
```

Reply:
The `useEffect` method won't actually work here, because the initialization happens *after* the first render, and because it's a ref the second render is never triggered.

The alertManager IS actually used by the rendered output, it just happens to be constant, so I think it's appropriate to use `useState` for this.  This may be slightly abusing `useState`, but it shouldn't actually be harmful (since the update is never called), and it's much cleaner than trying to fineggle it into a mutable ref (refs are for mutable things, this isn't mutable).  That's my approach, at least - I've used it several places including in the `useStaticInput` hook internal to the runtime data service.

Reply:
Alternatively, you could use `useRef` with inline initialization.  The downside here is that a re-render will trigger the `makeAlertManager` function, but at least it won't update the reference:

```js
export const AlertsProvider = ({ children }: { children: React.ReactNode }) => {
    const [alerts, setAlerts] = useState<Alert[]>([])
    const alertsManagerRef = useRef<AlertManager>(makeAlertManager(setAlerts))

    return (
        <AlertsManagerContext.Provider value={alertsManagerRef.current}>
            <AlertsContext.Provider value={alerts}>
                {children}
            </AlertsContext.Provider>
        </AlertsManagerContext.Provider>
    )
}
```

I don't find this particularly nicer than the `useState` variant, since using a mutable value in render output is an anti-pattern that could lead to issues if not noticed in a refactor, but they both should work.

Reply:
Note the typescript error is exposing the bug in the `useEffect` + `useRef` version, because on first render the ref is set to `undefined`, then updated to a new instance of an `AlertManager` later.  This means the implicit type of the ref is `AlertManager | undefined` (and that's only if you specify the type - otherwise the assignment will fail because `ref.current = new AlertManager` is incorrect when the ref implicitly has type `undefined` 

Since the ref is `undefined` on first render, it is an invalid value to pass to `AlertsManagerContext.Provider`'s `value` prop, which expects only an AlertManager.

Reply:
I have read through the docs again, and I think `useState` is the best solution as it supports lazy initial state (I'll omit types for readability):

```js
const [alertsManager] = useState(() => makeAlertManager(setAlerts))
```

Reply:
Yep, agreed - this was the original approach I took here

Reply:
My proposal has a difference (and I think it's exactly what @HendrikThePendric was looking for):
You can pass functions to `useState` and it'll run them only once.

Reply:
Ah yeah you’re right - I used that initializer function initially but must have taken it out.  Necessary when adding the `id` counter 👍

Reply:
> Note the typescript error is exposing the bug in the `useEffect` + `useRef` version, because on first render the ref is set to `undefined`, then updated to a new instance of an `AlertManager` later. This means the implicit type of the ref is `AlertManager | undefined` (and that's only if you specify the type - otherwise the assignment will fail because `ref.current = new AlertManager` is incorrect when the ref implicitly has type `undefined`

Thanks for confirming this @amcgee 

This was very helpful, I'll go with the lazy initial state approach. Thanks @amcgee and @Mohammer5

Reply:
OT: If the async initialization hadn't been an issue, you could've provided `defaultAlertManager` (from the `AlertsManagerContext.ts`) as the initial value to the ref and the type error would've disappeared